csrc/preseg_passes/remove_bcast_squeeze.cpp

```diff
@@ -8,46 +8,296 @@
+      if (!ops.empty() && ops.back() == AxisOp::SQUEEZE) {
+        ops.back() = AxisOp::PRESERVE;
+      } else {
+        ops.push_back(AxisOp::BROADCAST);
       }
+      next_pos++;
+    }
+    if (prev_pos >= prev.size() || next_pos >= next.size()) {
+      NVF_ERROR(
+          prev_pos >= prev.size() && next_pos >= next.size(),
+          "Failed to align some ops");
+      break;
     }
+    // Now we have prev[prev_pos] != SQUEEZE and next[next_pos] != BROADCAST,
+    // which means prev[prev_pos] provides the next op with its input. So here
+    // we are actually composing these aligned ops.
+    if (prev[prev_pos] == AxisOp::PRESERVE) {
+      ops.push_back(next[next_pos]);
+    } else if (next[next_pos] == AxisOp::PRESERVE) {
+      ops.push_back(prev[prev_pos]);
+    } else {
+      NVF_ERROR(
+          next[next_pos] == AxisOp::SQUEEZE &&
+          prev[prev_pos] == AxisOp::BROADCAST);
+      // Skip this op, since it is a "broadcast then squeeze" pattern
+    }
+    prev_pos++;
+    next_pos++;
+  }
+  return ops;
+}
 
-    // step-2: find and remove squeeze + broadcast pattern
-    // before: Y = squeeze(X); Z = broadcast(Y);  M = someOp(Z)
-    // after : M = someOp(X)
-    // conditions: (1) broadcast and squeeze have the same dim flags
-    if (auto bcast = dynamic_cast<BroadcastOp*>(expr)) {
-      if (auto squeeze = dynamic_cast<SqueezeOp*>(bcast->in()->definition())) {
-        if (bcast->getBroadcastDimFlags() == squeeze->getSqueezeDimFlags()) {
-          ir_utils::replaceValInAllExprInputsAndFusionOutputs(
-              bcast->out(), squeeze->in());
+TensorView* maybeDoReplacement(TensorView* orig) {
+  Expr* second = orig->definition();
+  if (!isReplaceableExpr(second)) {
+    return orig;
+  }
+  Expr* first = second->input(0)->definition();
+  if (!isReplaceableExpr(first)) {
+    return orig;
+  }
+
+  AxisOps first_ops = exprToAxisOps(first);
+  AxisOps second_ops = exprToAxisOps(second);
+  AxisOps simplified_ops = composeOps(first_ops, second_ops);
+  std::optional<AxisOp> simple_op_type_opt =
+      getSimplifiedOpType(simplified_ops);
+  if (!simple_op_type_opt.has_value()) {
+    return orig;
+  }
+  TensorView* replacement = orig;
+  if (simplified_ops == first_ops) {
+    // The second op was simply a "Set" operation, so we just skip it
+    replacement = first->output(0)->as<TensorView>();
+  } else {
+    TensorView* input_tv = first->input(0)->as<TensorView>();
+    switch (simple_op_type_opt.value()) {
+      case AxisOp::PRESERVE:
+        // This is equivalent to a set Op
+        replacement = input_tv;
+        // Check that parallelization is consistent for replacement. We do not
+        // want to alter the parallelization of input_tv here; we only
+        // parallelize _new_ TVs. If the parallelization is inconsistent, we
+        // simply refuse to make the replacement.
+        if (!hasCompatibleParallelization(orig, replacement)) {
```

Comment:
We do not. Previously I was not performing this check inside `isSimpleTVset`. Thanks for pointing this out.

Reply:
Actually, I do need to keep it here. If we simplify a "broadcast then squeeze" or "squeeze then broadcast" we might actually hit this path. In that case, it's possible that the input to the first op has different parallelization than the output to the second op. I assume we should skip the replacement in that setting...

Reply:
Thanks to @wujingyue's tip, I cleaned this up in d928305c by removing this check, and removing `hasCompatibleParallelization`. In `isReplaceableExpr`, I check `isResharding(expr)` which is a clearer way to avoid replacing reshardings.